src/lib/auth/auth0.ts

```diff
@@ -1,26 +1,242 @@
-  authParams: CustomAuthParams,
-): Promise<GetTokenSilentlyVerboseResponse> {
-  return auth0.getTokenSilently({
-    detailedResponse: true,
-    authorizationParams: authParams,
-    cacheMode: "off",
-  });
+import { toBuffer } from "../utils/bufferUtils";
+import { jwtDecode } from "jwt-decode";
+import {
+  CryptoOperationData,
+  AuthorizationParams,
+  AuthorizationParamsWithTransactionInput,
+  Auth0Strategy,
+  IdTokenWithData,
+  UserDetails,
+} from "./auth0.types";
+import { DEFAULT_REFRESH_TOKEN_EXPIRATION_MS } from "../config/config.constants";
+
+export class OthentAuth0Client {
+  auth0ClientPromise: Promise<Auth0Client | null> = Promise.resolve(null);
+
+  isInitialized = false;
+
+  userDetails: UserDetails | null = null;
+
+  static isUserValid<D>(
+    idTokenOrUser: IdTokenWithData<D> | UserDetails,
+  ): boolean {
+    // Note that we are not using the ID Token `exp` field, which is typically 24 hours. We don't care about that value
+    // as refresh tokens have a much longer expiration, 15 days typically.
+
+    const now = Date.now();
+
+    const expiration =
+      (idTokenOrUser as UserDetails).expiration ||
+      now + DEFAULT_REFRESH_TOKEN_EXPIRATION_MS;
+
+    return !!(
+      idTokenOrUser &&
+      idTokenOrUser.sub &&
+      idTokenOrUser.owner &&
+      idTokenOrUser.walletAddress &&
+      idTokenOrUser.authSystem === "KMS" &&
+      expiration > now
+    );
+  }
+
+  static getUserDetails<D>(idToken: IdTokenWithData<D>): UserDetails {
+    return {
+      sub: idToken.sub || "",
+      name: idToken.name,
+      givenName: idToken.given_name,
+      familyName: idToken.family_name,
+      nickname: idToken.nickname,
+      picture: idToken.picture,
+      locale: idToken.locale,
+      email: idToken.email,
+      emailVerified: idToken.email_verified,
+      expiration: Date.now() + DEFAULT_REFRESH_TOKEN_EXPIRATION_MS,
+      owner: idToken.owner,
+      walletAddress: idToken.walletAddress,
+      authSystem: idToken.authSystem,
+    };
+  }
+
+  static getAuthorizationParams(
+    authorizationParams?: AuthorizationParams,
+  ): AuthorizationParamsWithTransactionInput;
+  static getAuthorizationParams(
+    data?: CryptoOperationData,
+  ): AuthorizationParamsWithTransactionInput;
+  static getAuthorizationParams(
+    authorizationParamsOrData: AuthorizationParams | CryptoOperationData = {},
+  ): AuthorizationParamsWithTransactionInput {
+    const { authorizationParams, data } =
+      authorizationParamsOrData.hasOwnProperty("keyName")
+        ? {
+            authorizationParams: null,
+            data: authorizationParamsOrData as CryptoOperationData,
+          }
+        : {
+            authorizationParams:
+              authorizationParamsOrData as AuthorizationParams,
+            data: null,
+          };
+
+    return {
+      ...authorizationParams,
+      // TODO: We probably want to include the SDK and the API version here as we might want to deprecate the old one once
```

Comment:
See here. Including the SDK and API version in `transaction_input` might be useful.